client/trino-client/src/main/java/okhttp3/internal/tls/LegacyHostnameVerifier.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3.internal.tls;
+
+import javax.net.ssl.HostnameVerifier;
+import javax.net.ssl.SSLException;
+import javax.net.ssl.SSLSession;
+import javax.security.auth.x500.X500Principal;
+
+import java.security.cert.Certificate;
+import java.security.cert.X509Certificate;
+
+import static okhttp3.internal.Util.verifyAsIpAddress;
+import static okhttp3.internal.tls.OkHostnameVerifier.allSubjectAltNames;
+
+public class LegacyHostnameVerifier
+        implements HostnameVerifier
+{
+    public static final HostnameVerifier INSTANCE = new LegacyHostnameVerifier();
+
+    private LegacyHostnameVerifier() {}
+
+    @Override
+    public boolean verify(String host, SSLSession session)
+    {
+        if (OkHostnameVerifier.INSTANCE.verify(host, session)) {
+            return true;
+        }
+
+        // the CN cannot be used with IP addresses
+        if (verifyAsIpAddress(host)) {
+            return false;
+        }
+
+        // try to verify using the legacy CN rules
```

Comment:
Is this code copied from previous version of OkHttp client or did you wrote it yourself?

Reply:
It's a copy of the fallback code from the previous version. The only difference is that their code uses a private method to get just the `ALT_DNS_NAME`, whereas we use the public method `allSubjectAltNames()` which also includes `ALT_IPA_NAME`. This means if you have a certificate which has an alt name for an IP but not DNS, the old code would fallback to CN, but this will not. Such a certificate seems very unlikely, as it's a weird mix of new and legacy -- if you are using alt names for IP, you should also use them for DNS.